Apache Webtier support with nonroot user, non-privileged port and generic Dockerfile

ashageetha opened the pull request:
Hi @mriccell , @rjeberhard , @doxiao , 

Old PR oracle/docker-images#1630, got closed because of develop branch deletion. 
This PR is merged to master. Below are the details:

This PR has fix for Apache Webtier with below changes:

- Run as non-root ( oracle user)
- Support for both default privileged port (80) and non-privileged port (8080)
- Single Generic Dockerfile to build both 12.2.1.3 and 12.2.1.4 Apache Webtier image
- Readme updates 

Once this PR is merged then we can make the root directory to reflect the generic Dockerfile support by changing "12213-webtier-apache to webtier-apache".

This sample is being used as part of Apache Webtier Setup as per WebLogic Kubernetes Operator Samples: "https://github.com/oracle/weblogic-kubernetes-operator/tree/master/kubernetes/samples/charts/apache-webtier".
Hence the corresponding changes to  WebLogic Kubernetes Operator Samples are available in PR oracle/weblogic-kubernetes-operator#1916.

doxiao commented on OracleWebLogic/samples/12213-webtier-apache/Dockerfile:
```diff
@@ -46,16 +52,38 @@ COPY container-scripts/* /u01/oracle/container-scripts/
+    touch /etc/sudoers && \
+    sed -i /etc/sudoers -re 's/^%sudo.*/%sudo ALL=(ALL:ALL) NOPASSWD: ALL/g' && \
+    sed -i /etc/sudoers -re 's/^root.*/root ALL=(ALL:ALL) NOPASSWD: ALL/g' && \
+    echo "oracle ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers && \
```
Having looked into this more carefully, I believe the original line (below) between line#60 and #61 is still needed.
    sed -i /etc/sudoers -re 's/^#includedir.*/## **Removed the include directive** ##"/g' && \

ashageetha replied:
As per your original comment at "https://github.com/oracle/docker-images/pull/1630#discussion_r462495380", we removed this line and validated and it worked fine.
Please let me know if its still need to be added back.

doxiao replied:
Sorry for the confusion. I should have given more explanation in my comments.

First of all, either way works for your purposes in the sample in a simple environment. But there are differences.

— Removing the include directive (add back the origin line in the script) prevents accidental override of your new settings with any thing in the directory specified in that line. But it will also remove any additional settings in the includedir even if those setting do not impact your settings. 
— Keeping the include directive (as you have now in this PR) will allow any additional settings to take effect, but has the risk that your settings could be overridden by things in the includedir.
So I think it is slightly safer to remove the include directive, so I suggested to add that line back.

Djelibeybi replied:
Someone would have to modify the image manually in order to override this, so the risk of this occurring is very low. Keeping the includedir directive also maintains standard configuration (as per other images) so maintaining it is recommended so that the behaviour of this particular image matches other images in terms of how `sudo` is configured.

doxiao replied:
that's fine with me.

Djelibeybi replied:
In fact, now that I'm thinking about this more: instead of editing `/etc/sudoers`, you should just add a file to `/etc/sudoers.d` which will be included automatically by the default `includedir` line.

ashageetha replied:
Sure, we will try this option and verify.

ashageetha replied:
Please review latest commit with subject "Incorporated both comments from @doxiao and @Djelibeybi"

doxiao reviewed (commented):
@monica we probably should ask somebody (other than Asha) who has access to our OpenShit environment to test the changes and the corresponding changes in the WebLogic operator samples.

ashageetha commented on OracleWebLogic/samples/12213-webtier-apache/Dockerfile:
```diff
@@ -1,42 +1,48 @@
 # RUN:
-#      $ sh buildDockerImage.sh
+#      $ sh buildDockerImage.sh -v <version>
+#  The value of version is either 12.2.1.3 or 12.2.1.4   
```
Please review latest commit with subject "Incorporated both comments from @doxiao and @Djelibeybi"

Djelibeybi commented:
Looks ok to me from an OS perspective. Will let @mriccell and @doxiao review from the WLS perspective.

mriccell commented:
@rosemarymarano could you review?

doxiao commented:
@rosemarymarano  can you take a look att eh doc part of the changes in this PR? Thanks!

rosemarymarano commented on OracleWebLogic/samples/12213-webtier-apache/Dockerfile:
```diff
@@ -1,61 +1,92 @@
 # Example of Apache HTTP Server with WebLogic plugin for load balancing WebLogic on Docker Containers
 #
-# Copyright (c) 2016-2019 Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2016-2020 Oracle and/or its affiliates.
```
Copyright (c) 2016-2020 Oracle and/or its affiliates. -> Copyright (c) 2016, 2020, Oracle and/or its affiliates.

rosemarymarano commented on OracleWebLogic/samples/12213-webtier-apache/Dockerfile:
```diff
@@ -1,61 +1,92 @@
 # REQUIRED FILES TO BUILD THIS IMAGE
 # ----------------------------------
-# fmw_12.2.1.3.0_wlsplugins_Disk1_1of1.zip
+# For version 12.2.1.3 Required file fmw_12.2.1.3.0_wlsplugins_Disk1_1of1.zip
```
For version 12.2.1.3 Required -> For version 12.2.1.3, the required file is

rosemarymarano commented on OracleWebLogic/samples/12213-webtier-apache/Dockerfile:
```diff
@@ -1,61 +1,92 @@
 # ----------------------------------
-# fmw_12.2.1.3.0_wlsplugins_Disk1_1of1.zip
+# For version 12.2.1.3 Required file fmw_12.2.1.3.0_wlsplugins_Disk1_1of1.zip
+# For Version 12.2.1.4 Required file fmw_12.2.1.4.0_wlsplugins_Disk1_1of1.zip
```
For Version 12.2.1.4 Required file -> For version 12.2.1.4, the required file is

rosemarymarano commented on OracleWebLogic/samples/12213-webtier-apache/README.md:
```diff
@@ -1,42 +1,51 @@
 Apache HTTP Server with Oracle WebLogic Server Proxy Plugin on Docker
 ===============
-This project includes a quick start Dockerfile and samples for standalone Apache HTTP Server with the 12.2.1.3.0 Oracle WebLogic Server Proxy Plugin based on Oracle Linux. The certification of Apache on Docker does not require the use of any file presented in this repository. Customers and users are welcome to use them as starters, and customize, tweak, or create from scratch, new scripts and Doc ...
+This project includes a Dockerfile and samples for standalone Apache HTTP Server with the 12.2.1.3.0 and 12.2.1.4.0 Oracle WebLogic Server Proxy Plugin based on Oracle Linux. This is a generic Docker File and can be used to build apache images of different versions and to support containers to run on differnet ports i.e 80 and 8080. The certification of Apache on Docker does not require the use o ...
```
Docker File -> Dockerfile
apache -> Apache
differnet ports i.e 80 and 8080.->different ports, such as, 80 and 8080.
any file presented in this repository -> any file in this repository 
& -> and

rosemarymarano commented on OracleWebLogic/samples/12213-webtier-apache/README.md:
```diff
@@ -1,42 +1,51 @@
 ## Build Apache With the Plugin Docker Image
 
-This project offers a Dockerfile for the Apache HTTP Server with the Oracle WebLogic Server Proxy Plugin in standalone mode. To assist in building the images, you can use the  `buildDockerImage.sh` script. See below for instructions and usage.
+This project offers a generic Dockerfile for the Apache HTTP Server with the Oracle WebLogic Server Proxy Plugin in standalone mode to build different verions of the `Apache` image. To assist in building the images, you can use the  `buildDockerImage.sh` script. The only supported versions are 12.2.1.3.0 & 12.2.1.4.0. See below for instructions and usage.
```
verions  -> versions 
 & -> and

rosemarymarano commented on OracleWebLogic/samples/12213-webtier-apache/README.md:
```diff
@@ -1,42 +1,51 @@
+This project offers a generic Dockerfile for the Apache HTTP Server with the Oracle WebLogic Server Proxy Plugin in standalone mode to build different verions of the `Apache` image. To assist in building the images, you can use the  `buildDockerImage.sh` script. The only supported versions are 12.2.1.3.0 & 12.2.1.4.0. See below for instructions and usage.
 
-The `buildDockerImage.sh` script is a utility shell script that performs MD5 checks and is an easy way for beginners to get started. Expert users are welcome to directly call `docker build` with their preferred set of parameters.
+The `buildDockerImage.sh` script is a utility shell script that performs MD5 checks and construct the build args based on the verion of the apache image to be built, is an easy way for beginners to get started. Expert users are welcome to directly call `docker build` with their preferred set of parameters.
```
construct  -> constructs
verion  -> version 
apache  -> Apache 
built, is -> built; it is

rosemarymarano commented on OracleWebLogic/samples/12213-webtier-apache/README.md:
```diff
@@ -1,42 +1,51 @@
+The `buildDockerImage.sh` script is a utility shell script that performs MD5 checks and construct the build args based on the verion of the apache image to be built, is an easy way for beginners to get started. Expert users are welcome to directly call `docker build` with their preferred set of parameters.
 
-IMPORTANT: You have to download the `Oracle WebLogic Server Proxy Plugin 12.2.1.3.0` package (see the `.download` file) and place it in this directory.
+IMPORTANT: You have to download the `Oracle WebLogic Server Proxy Plugin` of version 12.2.1.3.0 and 12.2.1.4.0  package (see the `.download` file) and place it in this directory.
```
Plugin` of version 12.2.1.3.0 and 12.2.1.4.0  package -> Plugin`, version 12.2.1.3.0 or 12.2.1.4.0  package,

rosemarymarano commented on OracleWebLogic/samples/12213-webtier-apache/README.md:
```diff
@@ -52,39 +61,69 @@ NOTE: To set the `DOMAIN_NAME`, you must set both `DOMAIN_NAME` and `DOMAIN_HOME
+                     store/oracle/weblogic:<version>
 
-Start an Apache container by calling:
+Start an Apache container on default port by calling:
```
on default port -> on the default port

rosemarymarano commented on OracleWebLogic/samples/12213-webtier-apache/container-scripts/run-httpd.sh:
```diff
@@ -6,13 +6,19 @@
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
 #
-# Copyright (c) 2018 Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2018-2020 Oracle and/or its affiliates.
```
2018-2020 -> 2018, 2020,

rosemarymarano commented on OracleWebLogic/samples/12213-webtier-apache/container-scripts/run-httpd.sh:
```diff
@@ -6,13 +6,19 @@
 # Apache gets grumpy about PID files pre-existing
 rm -f /run/httpd/httpd.pid
 
+# Inorder to listen the apache container on port 8080, the variable NonPriviledgedPorts to be set to true.
```
Inorder to listen the apache -> To listen to the Apache
NonPriviledgedPorts to be set -> NonPriviledgedPorts must to be set

rosemarymarano commented on OracleWebLogic/samples/12213-webtier-apache/container-scripts/run-httpd.sh:
```diff
@@ -6,13 +6,19 @@
 rm -f /run/httpd/httpd.pid
 
+# Inorder to listen the apache container on port 8080, the variable NonPriviledgedPorts to be set to true.
+# Oracle user to update httpd conf file to support apache container on NonPriviledgedPorts 
```
apache  -> Apache

rosemarymarano commented on OracleWebLogic/samples/12213-webtier-apache/weblogic.conf:
```diff
@@ -1,7 +1,7 @@
-# Copyright (c) 2018 Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2018-2020 Oracle and/or its affiliates.
```
2018-2020 -> 2018, 2020,

rosemarymarano reviewed (commented):
Edits.

ashageetha commented:
Updated as per @rosemarymarano's review comments.  Based on the current comments on Copyright, updated buildDockerImage.sh as well. Please review the latest changes.

mriccell commented:
@xiancao

Djelibeybi commented:
@xiancao or @mriccell could one of you either approve or reject this pull request? It's been untouched for 20 days now.

ashageetha commented:
Hi @mriccell , @rjeberhard , @doxiao , @xiancao ,
Taken care of all comments but still this PR is not approved.
Please confirm if anything to be updated.

Thanks & Regards,
Asha

Djelibeybi commented:
It needs to be approved by @mriccell as she is the code owner. It can't be merged until that happens.

mriccell commented on OracleWebLogic/samples/12213-webtier-apache/README.md:
```diff
@@ -1,42 +1,51 @@
 
 ### Administration Server Only Example
 
 First, make sure that you have the WebLogic Server 12.2.1.3 install image. Pull the WebLogic install image from the DockerStore, `store/oracle/weblogic:12.2.1.3`, or build your own image, `oracle/weblogic:12.2.1.3-developer`, at [https://github.com/oracle/docker-images/tree/master/OracleWebLogic/dockerfiles/12.2.1.3](https://github.com/oracle/docker-images/tree/master/OracleWebLogic/dockerfiles/1 ...
```
WebLogic Server 12.2.1.3  and 12.2.1.4.  We do not use Docker Store any longer please change to indicate to pull image from Oracle Container Registry container-registry.oracle.com/middleware/weblogic:<version>

mriccell commented on OracleWebLogic/samples/12213-webtier-apache/README.md:
```diff
@@ -52,39 +61,69 @@ NOTE: To set the `DOMAIN_NAME`, you must set both `DOMAIN_NAME` and `DOMAIN_HOME
                      -e DOMAIN_NAME=abc_domain \
                      -p 7001:7001 \
-                     store/oracle/weblogic:12.2.1.3
+                     store/oracle/weblogic:<version>
```
container-registry.oracle.com/middleware/weblogic:<version>

ashageetha replied:
Updated the README.md as per your comments. Please confirm if the directory name "12213-webtier-apache" need to be changed now as now the Dockerfile supports both 12.2.1.3 and 12.2.1.4.
